### PR TITLE
feat: add SBC opcode

### PR DESCRIPTION
**Pull request by Benzammour**
Adds SBC opcode as well as unit tests

### CODE REVIEW

**Comment by mormod** [comment]
LGTM, but there are some things that should be addressed in separate PRs, such as defines for the single registers in dual register banks. Always typing `cpu.AF.bytes.high` cannot be the best way

File: test/alu_8bit_sbc_test.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+Test(sbc_a_n, sbc_a_d_res_only, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x9A;
+    const uint8_t a = rand() % (0xFF);
+    const uint8_t d = rand() % (0xFF);
+    const uint8_t expected = a - d - 1;
+    const uint16_t address = rand() % (0xFFFF);
+    set_flag(1, C_FLAG);
+
+    uint8_t actual = cpu_setup(opcode, address, a, d, &cpu.DE.bytes.high);
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:115]: I think that there are some flag checks missing here

File: test/alu_8bit_sbc_test.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+Test(sbc_a_n, sbc_a_e_res_only, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x9B;
+    const uint8_t a = rand() % (0xFF);
+    const uint8_t e = rand() % (0xFF);
+    const uint8_t expected = a - e - 1;
+    const uint16_t address = rand() % (0xFFFF);
+    set_flag(1, C_FLAG);
+
+    uint8_t actual = cpu_setup(opcode, address, a, e, &cpu.DE.bytes.low);
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:132]: Flag checks

File: test/alu_8bit_sbc_test.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+Test(sbc_a_n, sbc_a_h_res_only, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x9C;
+    const uint8_t a = rand() % (0xFF);
+    const uint8_t h = rand() % (0xFF);
+    const uint8_t expected = a - h - 1;
+    const uint16_t address = rand() % (0xFFFF);
+    set_flag(1, C_FLAG);
+
+    uint8_t actual = cpu_setup(opcode, address, a, h, &cpu.HL.bytes.high);
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:149]: Flag checks

File: test/alu_8bit_sbc_test.c [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+Test(sbc_a_n, sbc_a_l_res_only, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x9D;
+    const uint8_t a = rand() % (0xFF);
+    const uint8_t l = rand() % (0xFF);
+    const uint8_t expected = a - l - 1;
+    const uint16_t address = rand() % (0xFFFF);
+    set_flag(1, C_FLAG);
+
+    uint8_t actual = cpu_setup(opcode, address, a, l, &cpu.HL.bytes.low);
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:166]: Flag checks

File: test/alu_8bit_sbc_test.c [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    cpu.PC = address;
+    cpu.AF.bytes.high = a;
+    cpu.HL.word = word_address;
+    mmu_write_byte(address, opcode);
+    mmu_write_byte(word_address, num);
+
+    // do the actual emulation
+    cpu_step();
+    const uint8_t actual = cpu.AF.bytes.high;
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:194]: Flag checks

File: test/alu_8bit_sbc_test.c [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    // setup cpu
+    cpu.PC = address;
+    cpu.AF.bytes.high = a;
+    mmu_write_byte(address, opcode);
+    mmu_write_byte(address+1, num);
+
+    // do the actual emulation
+    cpu_step();
+    const uint8_t actual = cpu.AF.bytes.high;
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:220]: Flag checks

File: test/alu_8bit_sbc_test.c [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    const uint8_t expected = (uint8_t) (a - l - 1);
+    const uint16_t address = rand() % (0xFFFF);
+    set_flag(1, C_FLAG);
+
+    uint8_t actual = cpu_setup(opcode, address, a, l, &cpu.HL.bytes.low);
+
+    // C should be set because of borrow
+    // checks if integer promotion was successful
+    cr_expect(eq(u8, get_flag_bit(C_FLAG), 1));
+
```
Comment (marcluque) [inline test/alu_8bit_sbc_test.c:241]: Flag checks for other flags than C

File: CMakeLists.txt [hunk 8]
```diff
@@ -98,9 +98,10 @@ if (${TEST})
...
 
     set(TEST_FILES
-        test/common/util.c
+        src/cpu.c
         test/fixtures/cpu_mmu.c
-        test/alu_8bit_sub_test.c)
+        test/alu_8bit_sub_test.c
+        test/alu_8bit_sbc_test.c)
 
     set(TEST_SOURCES ${TEST_FILES} ${SRC_FILES})
```
Comment (marcluque) [inline CMakeLists.txt:106]: There's no need to include `src/cpu.c` in `TEST_FILES` since `TEST_SOURCES` includes `cpu.c`

File: src/cpu.h [hunk 9]
```diff
@@ -174,34 +177,121 @@ void OPC_ADD_A_d8(void);
...
+/**
+ * @brief Subtract the value stored in Register L from the value stored in A,
+ *        addtionally, subtract the carry flag, then store it in A.
+ */
+void OPC_SBC_A_L(void);
+
+/**
+ * @brief First fetches a byte from the address `HL`, subtracts the fetched byte from A
+ *        addtionally, subtract the carry flag, then store it in A.
+ */
```
Comment (marcluque) [inline src/cpu.h:284]: Maybe we could add a note here that `SBC A, (HL)` takes 8 cycles because it fetches a byte from memory? Same with `SUB A, (HL)`

File: src/cpu.h [hunk 10]
```diff
@@ -57,6 +57,9 @@ void cpu_step(void);
 __attribute__((pure)) uint16_t cpu_get_PC(void);
 void cpu_print_registers(void);
 
+__attribute__((pure)) uint8_t get_flag_bit(Flag f);
+void set_flag(uint8_t bit, Flag f);
```
Comment (marcluque) [inline src/cpu.h:61]: I think that both of these should be `always_inline` since that allows us to switch a `CALL` instruction with `AND` or something similar when optimization is < O2

File: test/alu_8bit_sub_test.c [hunk 11]
```diff
@@ -103,8 +102,8 @@ Test(sub_a_n, sub_a_c_borrow, .init = cpu_mmu_setup, .fini = cpu_teardown) {
 
 Test(sub_a_n, sub_a_d_res_only, .init = cpu_mmu_setup, .fini = cpu_teardown) {
     const uint8_t opcode = 0x92;
-    const uint8_t a = rand() % (0xFFFF);
-    const uint8_t d = rand() % (0xFFFF);
+    const uint8_t a = rand() % (0xFF);
```
Comment (marcluque) [inline test/alu_8bit_sub_test.c:105]: I think it would be a good idea to remove the `% (0xFF)`, etc.
See here: https://godbolt.org/z/ec6W7Mfhh

File: test/alu_8bit_sbc_test.c [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+    // check if PC is updated correctly
+    cr_expect(eq(u8, cpu.PC, address + 1));
+}
+
+Test(sbc_a_n, sbc_a_d_res_only, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x9A;
+    const uint8_t a = rand() % (0xFF);
+    const uint8_t d = rand() % (0xFF);
+    const uint8_t expected = a - d - 1;
```
[the remark this answers is not in the document]
Comment (Benzammour) [reply test/alu_8bit_sbc_test.c:110]: Sure thing, the intention was to try various values over time instead of the hardcoded values.
This also tests the Gameboy's ALU as we pass those numbers to the CPU.

File: test/alu_8bit_sbc_test.c (hunk 1 above)
Comment (Benzammour) [reply test/alu_8bit_sbc_test.c:115]: No, because the goal of the remaining tests is to simply check the correct value results from the arithmetic, as we checked the different flag configurations in the preceding test cases

File: test/alu_8bit_sbc_test.c (hunk 7 above)
Comment (Benzammour) [reply test/alu_8bit_sbc_test.c:241]: As the other flags have been tested in the first three unit tests, this is not the goal of this unit test.
Here we simply want to check for the integer promotion, which only includes the C flag in this specific edge case

File: test/alu_8bit_sbc_test.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (marcluque) [reply test/alu_8bit_sbc_test.c:110]: That's true 🤔 I suppose that a hybrid version is appropriate, i.e., test as many "edge cases" as possible "by hand", but supplement with these random instances.

File: test/alu_8bit_sbc_test.c (hunk 1 above)
Comment (marcluque) [reply test/alu_8bit_sbc_test.c:115]: Ouh I see now, hence the name of the test, sorry for the comments, I didn't get that. It makes sense since this is necessary because of the randomization

**Review by marcluque — APPROVED** [review]
